File: agent-driven-architecture-main/examples/typescript/src/blocks/escalation/contract.ts
// ── blocks/escalation/contract — the block's TRANSPORT (G12) ───────────────
// Two verbs, one reversible and one not; identical declarations otherwise.
// No Actor, no Authority, no Signature anywhere in this file (check C4) — the
// Command's `sig` is declared on the spine's parent, and a ToolResult never
// carries an actor at all, which is what makes one unforgeable upstream.

import type { CommandBase } from "@adr/spine/pure/command";
import type { EffectBase } from "@adr/spine/pure/effect";
import type { TicketId } from "@adr/spine/pure/ids";
import type { ToolResultBase } from "@adr/spine/pure/tool-result";
import { claims } from "@adr/spine/pure/tool-result";

// ── ToolResult cases ────────────────────────────────────────────────────────
export interface RequestEscalationResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "requestEscalation";
  readonly ticket: TicketId;
}

export interface ConfirmEscalationResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "confirmEscalation";
  readonly ticket: TicketId;
}

export type EscalationResult = RequestEscalationResult | ConfirmEscalationResult;

// ── Command cases ───────────────────────────────────────────────────────────
export interface RequestEscalationCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "requestEscalation";
  readonly ticket: TicketId;
}

export interface ConfirmEscalationCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "confirmEscalation";
  readonly ticket: TicketId;
}

export type EscalationCommand = RequestEscalationCommand | ConfirmEscalationCommand;

// ── Effect cases ────────────────────────────────────────────────────────────
/** IRREVERSIBLE. It fires only from the success branch of the confirm arm, and
 *  only after the boundary gate admitted the confirm. */
export interface PageOncall extends EffectBase {
  readonly kind: "PageOncall";
  /** IRREVERSIBLE: a human is woken. Declared as a LITERAL, so no arm can spell
   *  this leaf `Routine` (docs/DECISIONS.md:85). */
  readonly effectClass: "Irreversible";
  readonly ticket: TicketId;
}

export type EscalationEffect = PageOncall;

/** WHICH RESULTS THIS BLOCK'S ARM FOLDS, derived rather than written. The table
 *  is a mapped type over the union above (`spine/pure/tool-result`), so a case
 *  added there and a case claimed here are ONE edit: omit it and the property is
 *  missing, name a tool the union does not declare and the property is excess.
 *  The under-claiming predicate this used to be — green build, stale clause,
 *  fall-through at run time — is not writable in this form. */
export const isEscalationResult = claims<EscalationResult>({
  requestEscalation: true,
  confirmEscalation: true,
});

File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/blocks/escalation/Contract.kt
// ── blocks/escalation/contract — the block's transport ─────────────────────
// The gated slice: a REQUEST is reversible, a CONFIRM is not. Both are ordinary
// verbs with ordinary Command cases; the difference lives in the Verb table's
// reversibility classification and in the boundary gate, not in a second mechanic.

package adr.contract

import adr.spine.pure.CommandId
import adr.spine.pure.EffectClass
import adr.spine.pure.Signature
import adr.spine.pure.TicketId
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName

/**
 * A sealed CLASS extending the sealed CLASS ToolResult: `tool` is passed up the chain,
 * so every variant carries it by construction rather than by re-implementing it.
 */
sealed class EscalationResult(
    override val tool: ToolName,
    /**
     * G12 one level down: a BLOCK's sub-union declares its own shared property on its own
     * parent — IN THE CONSTRUCTOR, as `open val`, so the parent actually holds it. Every
     * escalation verb is about a ticket, so every variant carries one by construction,
     * and the block's fold arm never has to ask which case it has just to find out which
     * ticket it is talking about.
     */
    open val ticket: TicketId,
) : ToolResult(tool) {

    data class RequestEscalation(
        override val tool: ToolName,
        override val ticket: TicketId,
    ) : EscalationResult(tool, ticket)

    data class ConfirmEscalation(
        override val tool: ToolName,
        override val ticket: TicketId,
    ) : EscalationResult(tool, ticket)
}

/**
 * A sealed CLASS extending the sealed CLASS Command: tool/sig/id pass up the chain and
 * every variant carries authorship, permission and identity by construction (G12).
 */
sealed class EscalationCommand(
    override val tool: ToolName,
    override val sig: Signature,
    override val id: CommandId,
) : Command(tool, sig, id) {
    data class RequestEscalation(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val ticket: TicketId,
    ) : EscalationCommand(tool, sig, id)

    data class ConfirmEscalation(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val ticket: TicketId,
    ) : EscalationCommand(tool, sig, id)
}

sealed class EscalationEffect(
    override val at: Timestamp,
    effectClass: EffectClass,
) : Effect(at, effectClass) {
    /**
     * IRREVERSIBLE. Fires only inside the confirm arm's success branch — a human is
     * woken — so the class is fixed in the SUPERCLASS CALL below rather than in this
     * leaf's own constructor: no arm has an argument to launder it with, and
     * `copy()` covers only `(at, ticket)` (docs/DECISIONS.md:85). Gate check C17
     * denies constructing this leaf anywhere but the escalation arm.
     */
    data class PageOncall(
        override val at: Timestamp,
        val ticket: TicketId,
    ) : EscalationEffect(at, EffectClass.Irreversible)
}
